haystack/document_stores/pinecone.py

```diff
@@ -0,0 +1,682 @@
+        duplicate_documents = duplicate_documents or self.duplicate_documents
+        assert (
+            duplicate_documents in self.duplicate_documents_options
+        ), f"duplicate_documents parameter must be {', '.join(self.duplicate_documents_options)}"
+
+        if not self.pinecone_indexes.get(index):
+            self.pinecone_indexes[index] = self._create_index_if_not_exist(
+                vector_dim=self.vector_dim,
+                index=index,
+                metric_type=self.metric_type,
+                replicas=self.replicas,
+                shards=self.shards,
+            )
+
+        field_map = self._create_document_field_map()
+        document_objects = [Document.from_dict(d, field_map=field_map) if isinstance(d, dict) else d for d in documents]
+        document_objects = self._handle_duplicate_documents(
+            documents=document_objects, index=index, duplicate_documents=duplicate_documents
+        )
+        if len(document_objects) > 0:
+            add_vectors = False if document_objects[0].embedding is None else True
+            # I don't think below is required
+            with tqdm(
+                total=len(document_objects), disable=not self.progress_bar, position=0, desc="Writing Documents"
+            ) as progress_bar:
+                for i in range(0, len(document_objects), batch_size):
+                    ids = [doc.id for doc in document_objects[i : i + batch_size]]
+                    # metadata fields are stored in Pinecone
+                    metadata = [doc.meta for doc in document_objects[i : i + batch_size]]
+                    if add_vectors:
+                        embeddings = [doc.embedding for doc in document_objects[i : i + batch_size]]
+                        embeddings_to_index = np.array(embeddings, dtype="float32")
+
+                        if self.similarity == "cosine":
+                            self.normalize_embedding(embeddings_to_index)
+                        # to convert to list objects
+                        embeddings = [embed.tolist() for embed in embeddings]
+                        vectors = zip(ids, embeddings, metadata)
+                        self.pinecone_indexes[index].upsert(vectors=vectors)
+
+                    docs_to_write_in_sql = []
+                    for doc in document_objects[i : i + batch_size]:
+                        doc.meta["vector_id"] = doc.id
+                        docs_to_write_in_sql.append(doc)
+                    super(PineconeDocumentStore, self).write_documents(
+                        docs_to_write_in_sql, index=index, duplicate_documents=duplicate_documents
+                    )
+                    progress_bar.update(batch_size)
+            progress_bar.close()
+
+    def _create_document_field_map(self) -> Dict:
+        return {
+            self.index: self.embedding_field,
+        }
+
+    def _build_filter_clause(self, filters: Dict[str, Union[str, int, float, bool, list]]) -> dict:
```

Comment:
For uniformity with the remaining document stores, it would be better to add `convert_to_pinecone` method to the classes in `filter_utils.py`. Like this, we can simply call `LogicalFilterClause.parse(filters).convert_to_pinecone()`.

Reply:
Updated to use `filter_utils.py`, I've removed `_build_filter_clause`